plugins/hpi-cfu/fu-hpi-cfu-device.c

```diff
@@ -0,0 +1,1683 @@
+
+	if (!fu_usb_device_interrupt_transfer(FU_USB_DEVICE(self),
+					      END_POINT_ADDRESS,
+					      buf,
+					      sizeof(buf),
+					      &actual_length,
+					      FU_HPI_CFU_DEVICE_TIMEOUT,
+					      NULL,
+					      &error_local)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    error_local->message);
+		return FALSE;
+	}
+	fu_dump_raw(G_LOG_DOMAIN, "FirmwareUpdateOfferAccepted", buf, actual_length);
+
+	if (buf[13] == 0x01) {
+		g_debug("success");
+	} else {
+		if (buf[13] == 0x02) {
+			g_debug("offer accepted: %s", fu_cfu_rr_code_to_string(buf[9]));
+		} else {
+			g_debug("offer accepted: %s is not a reject",
+				fu_cfu_rr_code_to_string(buf[9]));
+		}
+	}
+
+	/* success */
+	*reply = buf[13];
+	return TRUE;
+}
+
+static gboolean
+fu_hpi_cfu_device_read_content_ack(FuHpiCfuDevice *self,
+				   gboolean *lastpacket,
+				   guint8 *report_id,
+				   gint *reason,
+				   guint8 *status,
+				   GError **error)
+{
+	gsize actual_length = 0;
+	guint8 buf[128] = {0};
+	g_autoptr(GError) error_local = NULL;
+
+	g_debug("sequence number: %d", self->sequence_number);
+	if (!fu_usb_device_interrupt_transfer(FU_USB_DEVICE(self),
+					      END_POINT_ADDRESS,
+					      buf,
+					      sizeof(buf),
+					      &actual_length,
+					      FU_HPI_CFU_DEVICE_TIMEOUT,
+					      NULL,
+					      &error_local)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    error_local->message);
+		return FALSE;
+	}
+	fu_dump_raw(G_LOG_DOMAIN, "ReadContentAck", buf, actual_length);
+
+	/* success */
+	*report_id = buf[0];
+	if (buf[0] == FIRMWARE_REPORT_ID) {
+		g_debug("status:%s response:%s",
+			fu_cfu_offer_status_to_string(buf[13]),
+			fu_cfu_rr_code_to_string(buf[9]));
+		if (buf[13] == 0x01) {
+			if (self->last_packet_sent == 1)
+				*lastpacket = TRUE;
+			g_debug("last_packet_sent:1");
+			*status = buf[13];
+		}
+		*status = buf[13];
+	} else {
+		g_debug("read_content_ack: buffer[5]: %02x, response:%s",
+			(guchar)buf[5],
+			fu_cfu_content_status_to_string(buf[5]));
+
+		if (buf[5] == 0x00) {
+			g_debug("read_content_ack:1");
+			if (self->last_packet_sent == 1)
+				*lastpacket = TRUE;
+			g_debug("last_packet_sent:%d", *lastpacket);
+			*status = buf[5];
+		} else {
+			*status = buf[5];
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_hpi_cfu_device_firmware_update_offer_rejected(gint8 reply)
+{
+	if (reply == FU_HPI_CFU_STATE_UPDATE_OFFER_REJECTED) {
+		g_debug("OfferRejected");
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static gboolean
+fu_hpi_cfu_device_send_end_offer_list(FuHpiCfuDevice *self, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+	guint8 buf[] = {
+	    0x25,
+	    0x02,
+	    0x00,
+	    0xff,
+	    0xa0,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	    0x00,
+	};
```

Comment:
can we have a helper structure for the format of the buffer instead of random bytes?  Then you can do something like:

```
struct hp_cfi_buf buf = { 0 };
buf.command = 0x25;
buf.length = 0x02;
```

etc

This will go a long way for readability and error catching for off by one etc.

Reply:
Furthermore it would be good to put such a structure in rustgen if you can.